src/metpy/xarray.py

```diff
@@ -1487,26 +1506,17 @@ def wrapper(*args, **kwargs):
         bound_args = signature(func).bind(*args, **kwargs)
         bound_args.apply_defaults()
 
-        # Search for DataArray in arguments
-        dataarray_arguments = [
-            value for value in bound_args.arguments.values()
-            if isinstance(value, xr.DataArray)
-        ]
-
         # Fill in vertical_dim
-        if (
-            len(dataarray_arguments) > 0
-            and 'vertical_dim' in bound_args.arguments
-        ):
-            try:
-                bound_args.arguments['vertical_dim'] = (
-                    dataarray_arguments[0].metpy.find_axis_number('vertical')
-                )
-            except AttributeError:
-                # If axis number not found, fall back to default but warn.
-                warnings.warn(
-                    'Vertical dimension number not found. Defaulting to initial dimension.'
-                )
+        if 'vertical_dim' in bound_args.arguments:
+            for a in dataarray_arguments(bound_args):
+                try:
+                    bound_args.arguments['vertical_dim'] = a.metpy.find_axis_number('vertical')
+                    break  # Only checking the first argument
+                except AttributeError:
+                    # If axis number not found, fall back to default but warn.
+                    warnings.warn(
+                        'Vertical dimension number not found. Defaulting to initial dimension.'
+                    )
```

Comment:
What was the motivation for changing to a loop here? I had implemented this to only go off of the first argument, since if the vertical axis wasn't there on the first argument, the whole thing would like fail due to inconsistencies in dimension ordering before and after broadcasting. But, I very well could have missed something you're correcting here.

If the loop behavior is the proper target, I think we should also change how the warning works (only warning after not finding the vertical dimension on any argument), and implement a test for the case where the vertical axis only shows up in later arguments.

Reply:
Well, the loop allows us to use `dataarray_arguments` as an iterator, which made the implementation, refactoring, and use simple. Previously in this function, we were actually assembling the entire list, but then only using the first argument anyway. With the break statement we're getting the same behavior of:
* Use the first argument if there is one
* Skip it if there are none

I'd really like to avoid assembling an entire list if we're not going to use the list. The other option is:
```python
try:
  arg = next(dataarray_arguments(bound_args))
except StopIteration:
  pass
```

Which seems worse.

Reply:
Oh, I guess the `break` in this case should go *after* the `AttributeError` handler. Also, I could learn the Python language properly--`next()` can have a default value.